### ISSUE TITLE
deterministic-zo test failure

### ISSUE THREAD

**Issue by samth** [racket/racket#3193]
Occasionally, the `tests/racket/deterministic-zo` test fails; here's an example on DrDr. http://drdr.racket-lang.org/54533/cs/pkgs/racket-test/tests/racket/deterministic-zo.rkt#output1

While debugging this, I saw the following (not exactly identical) failure.

```                                                                    
./racket/block.rkt                                                                                       
check-file: failed relative to built: #<path:/home/samth/sw/plt/racket/collects/racket> #<path:/home/samth/sw/plt/racket/collects/> #<path:./racket> #<path:block.rkt> #<path:/home/samth/sw/plt/racket/collects/racket/compiled/ta6le/block_rkt.zo>                                                                       
  context...:                                                                                            
   /home/samth/sw/plt/racket/src/build/cs/c/compiled/io.scm:34021:20: do-error                           
   /home/samth/sw/plt/pkgs/racket-test/tests/racket/deterministic-zo.rkt:47:2: for-loop                  
   [repeats 1 more time]                                                                                 
   /home/samth/sw/plt/pkgs/racket-test/tests/racket/deterministic-zo.rkt:61:0                            
   call-with-values                                                                                      
   call-in-empty-metacontinuation-frame                                                                  
   /home/samth/sw/plt/racket/src/build/cs/c/compiled/expander.scm:47099:121: body of "/home/samth/sw/plt/pkgs/racket-test/tests/racket/deterministic-zo.rkt"                                                      
   /home/samth/sw/plt/racket/src/build/cs/c/compiled/expander.scm:18640:43: for-loop                     
   /home/samth/sw/plt/racket/src/build/cs/c/compiled/expander.scm:18792:45: run-module-instance!         
   /home/samth/sw/plt/racket/src/build/cs/c/compiled/expander.scm:18560:31: for-loop                     
   /home/samth/sw/plt/racket/src/build/cs/c/compiled/expander.scm:18518:33: for-loop                     
   /home/samth/sw/plt/racket/src/build/cs/c/compiled/expander.scm:18771:37: run-module-instance!         
   .../commands/test.rkt:179:16                                                                          
   call-in-empty-metacontinuation-frame                                                                  
   call-with-empty-metacontinuation-frame-for-swap                                                       
deterministic-zo.rkt: racocs test: test raised an exception              
```

The c1 and c2 files generated for this error are attached, but they differ early on. Here's the beginning of each file:

```
[samth@homer:~/tmp plt] od -x c1 | head -2
0000000 7e23 3707 372e 302e 332e 630b 6568 2d7a
0000020 6373 6568 656d 9e42 a2ae 15a4 c606 0994
[samth@homer:~/tmp plt] od -x c2 | head -2
0000000 7e23 3707 372e 302e 332e 630b 6568 2d7a
0000020 6373 6568 656d b142 8ad6 2ca9 f17a 4cb7
```

[c.zip](https://github.com/racket/racket/files/4645368/c.zip)

**Comment by mflatt** [comment]
I tracked down one source of non-determinism that I could provoke on my machine, but I don't know whether it's related to other failures.

If there's any difference in ".zo" files, there will be some difference near the beginning, since a SHA-1 of the ".zo" content is close to the beginning of the file. The differences in "c.zip" look so large that I wonder whether it was drift in the compiler within a version, since I don't always bump the version number with changes that can affect the compiler's output. (The "deterministic-zo.rkt" test expects that everything is completely in sync.) I didn't rewind my installation to an older Chez Scheme version to investigate "c.zip" more.

**Comment by mflatt** [comment]
A failure after that fix: http://drdr.racket-lang.org/54552/cs/pkgs/racket-test/tests/racket/deterministic-zo.rkt

**Comment by mflatt** [comment]
The new patch is aimed at the DrDr failure linked above. Let's see whether it makes things any better.

**Comment by samth** [comment]
Here's a new error (it's possible this was fixed by racket/ChezScheme#af2f8784d96f09624b97476fb7b674a77097f9d0): http://drdr.racket-lang.org/54677/cs/pkgs/racket-test/tests/racket/deterministic-zo.rkt#output1

Zip file with c1 and c2 attached.
[fail.zip](https://github.com/racket/racket/files/4731896/fail.zip)

**Comment by samth** [comment]
Here's yet another one: http://drdr.racket-lang.org/54840/cs/pkgs/racket-test/tests/racket/deterministic-zo.rkt
[c.zip](https://github.com/racket/racket/files/4837372/c.zip)
